File: YuGiOh2/Cards/C23424603.cs
﻿using System.Linq;
using YuGiOh2.Base;

namespace YuGiOh2.Cards
{
    /// <summary>
    /// 荒野
    /// </summary>
    public class C23424603
    {
        public static int Type { get; } = (int)ChooseTargetType.None;

        public static bool CheckIfAvailable(Player player)
        {
            return true;
        }

        public static void ProcessEffect(Player player)
        {
            PowerUp(player);
            PowerUp(player.Enemy);
        }

        private static void PowerUp(Player player)
        {
            player.Field.MonsterFields
                .Where(c => c != null &&
                (c.MonsterType == (int)MonsterType.Dinosaur ||
                c.MonsterType == (int)MonsterType.Zombie ||
                c.MonsterType == (int)MonsterType.Rock))
                .ToList()
                .ForEach(c => { c.ATK += (c.ATK >> 1); c.DEF += (c.DEF >> 1); });
        }

        private static void PowerDown(Player player)
        {
            player.Field.MonsterFields
                .Where(c => c != null &&
                (c.MonsterType == (int)MonsterType.Dinosaur ||
                c.MonsterType == (int)MonsterType.Zombie ||
                c.MonsterType == (int)MonsterType.Rock))
                .ToList()
                .ForEach(c => { c.ATK = (int)(c.ATK / 1.5); c.DEF = (int)(c.DEF / 1.5); });
        }

        public static void ProcessWhenSummon(string targetID, Player player)
        {
            MonsterCard target = player.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID) ??
                player.Enemy.Field.MonsterFields.FirstOrDefault(c => c != null && c.UID == targetID);
            if (target == null ||
                (!(target.MonsterType == (int)MonsterType.Dinosaur ||
                target.MonsterType == (int)MonsterType.Zombie ||
                target.MonsterType == (int)MonsterType.Rock)))
                return;
            target.ATK += (target.ATK >> 1);
            target.DEF += (target.DEF >> 1);
        }

        public static void ProcessWhenSetMonster(string targetID, Player player)
        {
            ProcessWhenSummon(targetID, player);
        }

        public static void ProcessWhenLeave(string targetID, Player player)
        {
            PowerDown(player);
            PowerDown(player.Enemy);
        }
    }
}

File: YuGiOh2/Scripts/C23424603.lua
Type = 0;
--荒野
function checkIfAvailable(player)
    return true;
end

function processEffect(player)
    powerUp(player);
    powerUp(player.Enemy);
end

function powerUp(player)
    for index, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and (monster.MonsterType == 2 or monster.MonsterType == 10 or
                monster.MonsterType == 17) then
            monster.ATK = monster.ATK + monster.ATK / 2;
            monster.DEF = monster.DEF + monster.DEF / 2;
        end
    end
end

function powerDown(player)
    for index, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and (monster.MonsterType == 2 or monster.MonsterType == 10 or
                monster.MonsterType == 17) then
            monster.ATK = monster.ATK / 1.5;
            monster.DEF = monster.DEF / 1.5;
        end
    end
end

function processWhenSummon(targetID, player)
    local target = nil;
    for index, monster in pairs(player.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    for index, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and monster.UID == targetID then
            target = monster;
        end
    end
    if target ~= nil and (target.MonsterType == 2 or target.MonsterType == 10 or
            target.MonsterType == 17) then
        target.ATK = target.ATK + target.ATK / 2;
        target.DEF = target.DEF + target.DEF / 2;
    end
end

function processWhenSetMonster(targetID, player)
    processWhenSummon(targetID, player);
end

function processWhenSelfLeave(targetID, player)
    powerDown(player);
    powerDown(player.Enemy);
end
